simple/app.go

```diff
@@ -0,0 +1,334 @@
+	clientGenerator    resource.ClientGenerator
+	kinds              map[string]AppManagedKind
+	internalKinds      map[string]resource.Kind
+	cfg                AppConfig
+	converters         map[string]Converter
+}
+
+// AppConfig is the configuration used by App
+type AppConfig struct {
+	Name           string
+	KubeConfig     rest.Config
+	InformerConfig AppInformerConfig
+	ManagedKinds   []AppManagedKind
+	Converters     map[schema.GroupKind]Converter
+}
+
+// AppInformerConfig contains configuration for the App's internal operator.InformerController
+type AppInformerConfig struct {
+	ErrorHandler       func(context.Context, error)
+	RetryPolicy        operator.RetryPolicy
+	RetryDequeuePolicy operator.RetryDequeuePolicy
+	FinalizerSupplier  operator.FinalizerSupplier
+}
+
+// AppManagedKind is a Kind and associated functionality used by an App.
+type AppManagedKind struct {
+	// Kind is the resource.Kind being managed. This is equivalent to a kubernetes GroupVersionKind
+	Kind resource.Kind
+	// Reconciler is an optional reconciler to run for this Kind. Only one version of a Kind should have a Reconciler,
+	// otherwise, duplicate events will be received.
+	Reconciler operator.Reconciler
+	// Watcher is an optional Watcher to run for this Kind. Only one version of a Kind should have a Watcher,
+	// otherwise, duplicate events will be received.
+	Watcher operator.ResourceWatcher
+	// Validator is an optional ValidatingAdmissionController for the Kind. It will be run only for validation
+	// of this specific version.
+	Validator resource.ValidatingAdmissionController
+	// Mutator is an optional MutatingAdmissionController for the Kind. It will be run only for mutation
+	// of this specific version.
+	Mutator resource.MutatingAdmissionController
+	// SubresourceRoutes are an optional map of subresource paths to a route handler.
+	// If supported by the runner, calls to these subresources on this particular version will call this handler.
+	SubresourceRoutes map[string]func(ctx context.Context, writer http.ResponseWriter, req *app.SubresourceRequest) error
+	// ReconcileOptions are the options to use for running the Reconciler or Watcher for the Kind, if one exists.
+	ReconcileOptions ReconcileOptions
+}
+
+// ReconcileOptions are settings for the ListWatch and informer setup for a reconciliation loop
+type ReconcileOptions struct {
+	// Namespace is the namespace to use in the ListWatch request
+	Namespace string
+	// LabelFilters are any label filters to apply to the ListWatch request
+	LabelFilters []string
+	// FieldSelectors are any field selector filters to apply to the ListWatch request
+	FieldSelectors []string
+	// UsePlain can be set to true to avoid wrapping the Reconciler or Watcher in its Opinionated variant.
+	UsePlain bool
+}
+
+// NewApp creates a new instance of App, managing the kinds provided in AppConfig.ManagedKinds.
+// AppConfig MUST contain a valid KubeConfig to be valid. Kinds can be managed by the app either with
```

Comment:
Can we validate it and return an error if validation fails?

Reply:
We can validate that it is present, but I'm not sure what to check to see if it's valid, aside from attempting to use it to talk to a known resource. I suppose we could do thatusing a resource from the manifest. Thoughts?

Reply:
I left another comment – I thought there was a client-side only validation function in the client-go but turns out I was wrong. Let's maybe write something like that but it can be done in a follow-up PR(s), later on.